File: app/Console/Commands/AgiServer.php
<?php

namespace App\Console\Commands;

use App\Services\Agi\AgiConnection;
use App\Services\Agi\CallEndHandler;
use App\Services\Agi\InboundCallHandler;
use App\Services\Agi\OutboundCallHandler;
use Illuminate\Console\Command;
use Illuminate\Support\Facades\Log;

class AgiServer extends Command
{
    protected $signature = 'agi:serve {--host=0.0.0.0} {--port=4573}';

    protected $description = 'Start the FastAGI routing server for Asterisk';

    public function handle(): int
    {
        $host = $this->option('host');
        $port = $this->option('port');
        $address = "tcp://{$host}:{$port}";

        $server = @stream_socket_server($address, $errno, $errstr);

        if (!$server) {
            $this->error("Failed to start AGI server on {$address}: {$errstr} ({$errno})");
            return self::FAILURE;
        }

        $this->info("FastAGI server listening on {$address}");
        Log::info("FastAGI server started on {$address}");

        stream_set_blocking($server, false);

        $running = true;

        if (function_exists('pcntl_signal')) {
            pcntl_signal(SIGTERM, function () use (&$running) {
                $running = false;
            });
            pcntl_signal(SIGINT, function () use (&$running) {
                $running = false;
            });
        }

        while ($running) {
            if (function_exists('pcntl_signal_dispatch')) {
                pcntl_signal_dispatch();
            }

            $read = [$server];
            $write = null;
            $except = null;

            if (stream_select($read, $write, $except, 1) > 0) {
                $client = @stream_socket_accept($server, 5);

                if ($client) {
                    $this->handleConnection($client);
                }
            }
        }

        fclose($server);
        $this->info('FastAGI server stopped.');

        return self::SUCCESS;
    }

    private function handleConnection($socket): void
    {
        try {
            $agi = new AgiConnection($socket);
            $agi->parse();

            $script = $agi->getScript();

            Log::debug("AGI request: {$script}", [
                'channel' => $agi->getEnv('channel'),
                'extension' => $agi->getEnv('extension'),
                'callerid' => $agi->getEnv('callerid'),
            ]);

            match ($script) {
                'route_outbound' => app(OutboundCallHandler::class)->handle($agi),
                'route_inbound' => app(InboundCallHandler::class)->handle($agi),
                'call_end' => app(CallEndHandler::class)->handle($agi),
                default => $agi->verbose("rSwitch: Unknown AGI script: {$script}", 1),
            };
        } catch (\Throwable $e) {
            Log::error("AGI error: {$e->getMessage()}", [
                'trace' => $e->getTraceAsString(),
            ]);
        } finally {
            if (is_resource($socket)) {
                fclose($socket);
            }
        }
    }
}

File: python-services/call_control/agi_server.py
"""
FastAGI Server — Port of app/Console/Commands/AgiServer.php

Async TCP server that handles AGI requests from Asterisk.
Routes to appropriate handler based on script name in AGI request URL.

Scripts:
- route_outbound → OutboundCallHandler
- route_inbound  → InboundCallHandler
- call_end       → CallEndHandler
- broadcast_call → BroadcastCallHandler
- forward_call   → ForwardCallHandler
"""

import asyncio
import logging

from call_control.agi_protocol import AgiConnection
from call_control.outbound_handler import OutboundCallHandler
from call_control.inbound_handler import InboundCallHandler
from call_control.call_end_handler import CallEndHandler
from call_control.broadcast_handler import BroadcastCallHandler
from call_control.forward_handler import ForwardCallHandler

logger = logging.getLogger(__name__)

# Handler instances
_outbound = OutboundCallHandler()
_inbound = InboundCallHandler()
_call_end = CallEndHandler()
_broadcast = BroadcastCallHandler()
_forward = ForwardCallHandler()


async def handle_connection(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    """Handle a single AGI connection from Asterisk.

    No DB session is opened here — handlers offload sync DB work via
    shared.database.db_thread(), which checks out a fresh session per
    transactional unit. This keeps the asyncio event loop responsive
    even under sustained 50-70 cps load.
    """
    peer = writer.get_extra_info("peername")
    conn = AgiConnection(reader, writer)

    try:
        await conn.parse()

        script = conn.get_script()
        logger.info(f"AGI request: script={script}, channel={conn.get_channel()}")

        if script == "route_outbound":
            await _outbound.handle(conn)
        elif script == "route_inbound":
            await _inbound.handle(conn)
        elif script == "call_end":
            await _call_end.handle(conn)
        elif script == "broadcast_call":
            await _broadcast.handle(conn)
        elif script == "forward_call":
            await _forward.handle(conn)
        else:
            logger.warning(f"Unknown AGI script: {script}")
            await conn.verbose(f"rSwitch: Unknown script '{script}'")

    except Exception as e:
        logger.error(f"AGI connection error from {peer}: {e}", exc_info=True)
    finally:
        conn.close()


async def start_agi_server(host: str = "0.0.0.0", port: int = 4573):
    """Start the FastAGI TCP server."""
    try:
        server = await asyncio.start_server(handle_connection, host, port)
        addr = server.sockets[0].getsockname()
        logger.info(f"FastAGI server listening on {addr[0]}:{addr[1]}")

        async with server:
            await server.serve_forever()
    except OSError as e:
        if e.errno == 98:
            # Another worker already bound this port — that's OK
            logger.info(f"FastAGI port {port} already in use (another worker has it)")
        else:
            logger.error(f"FastAGI server failed to start: {e}")
